SchemeBoard backup / restore backend + stub frontend pages

jepett0 opened the pull request:
### Changelog entry <!-- a user-readable short description of the changes that goes to CHANGELOG.md and Release Notes -->

### Changelog category <!-- remove all except one -->

* Not for changelog (changelog entry is not required)

### Description for reviewers <!-- (optional) description for those who read this PR -->

Epic:
- https://github.com/ydb-platform/ydb/issues/20106

Issues:
- https://github.com/ydb-platform/ydb/issues/23975
- https://github.com/ydb-platform/ydb/issues/23976
- https://github.com/ydb-platform/ydb/issues/23978

jepett0 committed:
Add Scheme Board recovery monitoring page

- allows to backup and restore the scheme board replicas state
- gets the path descriptions from replicas by traversing them from the root to the leaves
- uploads the path descriptions on restore via a populator
- tracks progress

jepett0 committed:
remove frontend

CyberROFL reviewed (changes requested):
Новые акторы и вспомогательные классы лучше вынести в scheme_board_backup.cpp

jepett0 committed:
review fixes: 1

jepett0 committed:
templated progress handlers

jepett0 committed:
Update ydb/core/tx/scheme_board/monitoring.cpp

Co-authored-by: Ilnaz Nizametdinov <i.nizametdinov@gmail.com>

jepett0 committed:
Update ydb/core/tx/scheme_board/monitoring.cpp

Co-authored-by: Ilnaz Nizametdinov <i.nizametdinov@gmail.com>

jepett0 committed:
review fixes: extra diff, batch 3

jepett0 committed:
extrapolate review fixes

jepett0 committed:
review fix 4: remove explicit progress member

jepett0 commented on ydb/core/tx/scheme_board/monitoring.cpp:
```diff
@@ -1055,8 +1574,44 @@ class TMonitoring: public TActorBootstrapped<TMonitoring> {
+        bool backupStarted = false,
+        const TString& errorMessage = ""
+    ) {
+        Y_UNUSED(backupProgress, backupStarted, errorMessage);
```
да. На самом деле frontend спрятан уже в этом PR (в виде удалённого кода во втором коммите). Они понадобятся, потому что прогресс будет "персистентным", он должен переживать обновление страницы и я выбрал хранить его в TMonitoring акторе. Насчёт backupStarted и errorMessage можно подумать. Кажется, что их можно из backupProgress получить

jepett0 commented on ydb/core/tx/scheme_board/monitoring.cpp:
```diff
@@ -40,8 +49,477 @@ namespace NSchemeBoard {
+
+        if (!descriptions.empty()) {
+            Populate(std::move(descriptions));
+        }
```
да

CyberROFL replied:
Может хотя бы ошибку какую вернуть? Иначе совершенно непонятно, что произошло, даже не логируется ничего.

jepett0 committed:
Apply suggestions from code review

Co-authored-by: Ilnaz Nizametdinov <i.nizametdinov@gmail.com>

jepett0 committed:
review fixes, batch 5: const + style

jepett0 committed:
move new actors to a separate file

jepett0 committed:
compilation fix

jepett0 committed:
remove backup proxy actor

- use cookies to match events with the paths they correspond to

CyberROFL commented on ydb/core/tx/scheme_board/monitoring.cpp:
```diff
@@ -1294,18 +1383,137 @@ class TMonitoring: public TActorBootstrapped<TMonitoring> {
+                    ));
+                }
+
+                TString filePath = params.Get("backupPath");
```
```suggestion
                const TString filePath = params.Get("backupPath");
```

CyberROFL commented on ydb/core/tx/scheme_board/monitoring.cpp:
```diff
@@ -1294,18 +1383,137 @@ class TMonitoring: public TActorBootstrapped<TMonitoring> {
+                    ));
+                }
+
+                TString filePath = params.Get("restorePath");
```
```suggestion
                const TString filePath = params.Get("restorePath");
```

CyberROFL commented on ydb/core/tx/scheme_board/monitoring.cpp:
```diff
@@ -1294,18 +1383,137 @@ class TMonitoring: public TActorBootstrapped<TMonitoring> {
         Send(ev->Sender, new NMon::TEvHttpInfoRes(NMonitoring::HTTPNOTFOUND, 0, EContentType::Custom));
     }
 
+    template <typename TProgress, typename TEvent>
+    void Handle(TProgress& progress, TEvent::TPtr& ev) {
```
```suggestion
    template <typename TProgress, typename TEventPtr>
    void Handle(TProgress& progress, TEventPtr& ev) {
```

Тогда `TEventPtr` будет выводится автоматически.

CyberROFL commented on ydb/core/tx/scheme_board/backup.h:
```diff
@@ -0,0 +1,105 @@
+#include "helpers.h"
```
```suggestion
#pragma once

#include "helpers.h"
```

CyberROFL commented on ydb/core/tx/scheme_board/backup.h:
```diff
@@ -0,0 +1,105 @@
+
+    TBackupProgress() = default;
+
+    explicit TBackupProgress(const TSchemeBoardMonEvents::TEvBackupProgress& ev)
```
Рекомендую сделать forward declaration и унести тело конструктора в cpp-файл.

CyberROFL commented on ydb/core/tx/scheme_board/backup.h:
```diff
@@ -0,0 +1,105 @@
+};
+
+struct TRestoreProgress {
+    enum class EStatus {
+        Idle,
+        Starting,
+        Running,
+        Completed,
+        Error,
+    };
+
+    ui32 TotalPaths = 0;
+    ui32 ProcessedPaths = 0;
+    EStatus Status = EStatus::Idle;
+    TString ErrorMessage;
```
Судя по тому, что всё это повторяется для Backup и Restore, можно сделать `TCommonProgress`.

CyberROFL commented on ydb/core/tx/scheme_board/backup.cpp:
```diff
@@ -0,0 +1,394 @@
+            const TString path = PendingPaths.front();
+            PendingPaths.pop();
+
+            const ui64 cookie = PathByCookie.size();
```
Для формирования cookie лучше сделать отдельный счетчик. `PathByCookie` из вектора преобразовать в `THashMap<ui64, TString>` и удалять из него ненужные элементы, а размер словаря использовать для контроля инфлайта.

CyberROFL commented on ydb/core/tx/scheme_board/backup.cpp:
```diff
@@ -0,0 +1,394 @@
+            // parse children and add to pending queue
+            try {
+                TJsonValue parsedJson;
+                if (NJson::ReadJsonFastTree(jsonDescription, &parsedJson)) {
```
Зачем сериализовывать в json, а потом обратно парсить, если можно из исходного протобуфа достать список детей?

CyberROFL commented on ydb/core/tx/scheme_board/backup.cpp:
```diff
@@ -0,0 +1,394 @@
+            TTwoPartDescription description;
+            auto status = google::protobuf::json::JsonStringToMessage(line, &description.Record);
+            if (!status.ok()) {
+                return ReplyError(TStringBuilder() << "Failed to parse JSON line: " << TStringBuf(line, 0, 100) << ", status: " << status.ToString());
```
```suggestion
                return ReplyError(TStringBuilder() << "Failed to parse JSON"
                    << ": line: " << TStringBuf(line, 0, 100)
                    << ", status: " << status.ToString());
```

CyberROFL commented on ydb/core/tx/scheme_board/backup.cpp:
```diff
@@ -0,0 +1,394 @@
+
+        while (InputFile->ReadLine(line)) {
+            TTwoPartDescription description;
+            auto status = google::protobuf::json::JsonStringToMessage(line, &description.Record);
```
```suggestion
            const auto status = google::protobuf::json::JsonStringToMessage(line, &description.Record);
```

CyberROFL commented on ydb/core/tx/scheme_board/backup.cpp:
```diff
@@ -0,0 +1,394 @@
+            if (!status.ok()) {
+                return ReplyError(TStringBuilder() << "Failed to parse JSON line: " << TStringBuf(line, 0, 100) << ", status: " << status.ToString());
+            }
+            TPathId pathId(description.Record.GetPathOwnerId(), description.Record.GetPathId());
```
```suggestion
            const TPathId pathId(description.Record.GetPathOwnerId(), description.Record.GetPathId());
```

CyberROFL commented on ydb/core/tx/scheme_board/backup.cpp:
```diff
@@ -0,0 +1,394 @@
+            SBB_LOG_N("Unexpected info response");
+        }
+        const auto& info = ev->Get()->Record.GetPopulatorResponse();
+        TPathId maxRequestedPathId(info.GetMaxRequestedPathId().GetOwnerId(), info.GetMaxRequestedPathId().GetLocalPathId());
```
```suggestion
        const auto maxRequestedPathId = TPathId::FromProto(info.GetMaxRequestedPathId());
```

CyberROFL replied:
Давай тогда отдельно (до или после) выпилим из scheme_board_mon.proto кастомный `TPathId`.

CyberROFL commented on ydb/core/tx/scheme_board/backup.cpp:
```diff
@@ -0,0 +1,394 @@
+        SBB_LOG_D("Handle " << ev->Get()->ToString());
+        if (ev->Sender != Populator || !ev->Get()->Record.HasPopulatorResponse()) {
+            SBB_LOG_N("Unexpected info response");
+        }
+        const auto& info = ev->Get()->Record.GetPopulatorResponse();
```
```suggestion
        }

        const auto& info = ev->Get()->Record.GetPopulatorResponse();
```

CyberROFL commented on ydb/core/tx/scheme_board/backup.cpp:
```diff
@@ -0,0 +1,394 @@
+        SendProgressUpdate();
+        if (ProcessedPaths == TotalPaths) {
+            return ReplySuccess();
+        }
+        TActivationContext::Schedule(ProgressPollingInterval, new IEventHandle(Populator, SelfId(), new TSchemeBoardMonEvents::TEvInfoRequest(1)));
```
```suggestion
        }

        TActivationContext::Schedule(ProgressPollingInterval, new IEventHandle(Populator, SelfId(), new TSchemeBoardMonEvents::TEvInfoRequest(1)));
```

CyberROFL commented on ydb/core/tx/scheme_board/backup.cpp:
```diff
@@ -0,0 +1,394 @@
+    }
+
+private:
+    void ReadAndProcessFile() {
```
Т.к. читаем весь файл за один цикл обработки, то про это тоже стоит куда-нибудь написать (плашку?). Что актор-система на ноде будет заблочена обработкой файла резервной копии.

jepett0 committed:
Apply suggestions from code review

Co-authored-by: Ilnaz Nizametdinov <i.nizametdinov@gmail.com>

jepett0 committed:
tiny style fix

jepett0 committed:
compilation fix after applying suggestions

jepett0 committed:
error on empty descriptions

jepett0 committed:
communism for progress and result events and holders

jepett0 committed:
move progress constructors definitions to .cpp

jepett0 committed:
hash map to store path by cookie mapping

jepett0 commented on ydb/core/tx/scheme_board/monitoring.cpp:
```diff
@@ -1294,18 +1383,136 @@ class TMonitoring: public TActorBootstrapped<TMonitoring> {
+        progress = TProgress(msg);
+    }
+
+    void Handle(TSchemeBoardMonEvents::TEvBackupProgress::TPtr& ev) {
+        Handle<TBackupProgress, TSchemeBoardMonEvents::TEvBackupProgress::TPtr>(BackupProgress, ev);
+    }
+
+    void Handle(TSchemeBoardMonEvents::TEvBackupResult::TPtr& ev) {
+        Handle<TBackupProgress, TSchemeBoardMonEvents::TEvBackupResult::TPtr>(BackupProgress, ev);
+    }
+
+    void Handle(TSchemeBoardMonEvents::TEvRestoreProgress::TPtr& ev) {
+        Handle<TRestoreProgress, TSchemeBoardMonEvents::TEvRestoreProgress::TPtr>(RestoreProgress, ev);
+    }
+
+    void Handle(TSchemeBoardMonEvents::TEvRestoreResult::TPtr& ev) {
+        Handle<TRestoreProgress, TSchemeBoardMonEvents::TEvRestoreResult::TPtr>(RestoreProgress, ev);
+    }
```
работает, поправлю

jepett0 commented on ydb/core/tx/scheme_board/backup.cpp:
```diff
@@ -0,0 +1,395 @@
+        --InProgressPaths;
+    }
+
+    void HandleTimeout() {
+        SBB_LOG_D("Timeout");
+        --InProgressPaths;
+    }
+
+    void HandleUndelivered() {
+        SBB_LOG_D("Undelivered");
+        --InProgressPaths;
+    }
```
А откуда нам его здесь взять?

jepett0 committed:
Apply suggestions from code review

Co-authored-by: Ilnaz Nizametdinov <i.nizametdinov@gmail.com>

jepett0 committed:
track undelivered and timeouts path by path

jepett0 committed:
tiny style fix to minimize diff

jepett0 committed:
forward declare mon events

It is not possible to just forward declare the inner structs, so I have moved them outside the TSchemeBoardMonEvents struct.

jepett0 committed:
remove duplicate paths in progress counter

jepett0 commented on ydb/core/tx/scheme_board/backup.cpp:
```diff
@@ -0,0 +1,410 @@
+            return;
+        }
+        SBB_LOG_I("Timeout");
+        MarkPathCompleted(it);
```
сделаю отдельно, записал в follow-ups:
- https://github.com/ydb-platform/ydb/issues/24596
